source/extensions/transport_sockets/tls/cert_validator/default_validator.cc

```diff
@@ -276,6 +303,86 @@ Envoy::Ssl::ClientValidationStatus DefaultCertValidator::verifyCertificate(
+    *out_alert = SSL_AD_CERTIFICATE_UNKNOWN;
+  }
+  if (sk_X509_num(&cert_chain) == 0) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(
+          Envoy::Ssl::ClientValidationStatus::NotValidated);
+    }
+    stats_.fail_verify_error_.inc();
+    std::string error("verify cert failed: empty cert chain");
+    if (error_details != nullptr) {
+      *error_details = error;
+    }
+    ENVOY_LOG(debug, error);
+    return Ssl::ValidateResult::Failed;
+  }
+  X509* leaf_cert = sk_X509_value(&cert_chain, 0);
+  ASSERT(leaf_cert);
+  if (verify_trusted_ca_) {
+    X509_STORE* verify_store = SSL_CTX_get_cert_store(ssl_ctx);
+    ASSERT(verify_store);
+    bssl::UniquePtr<X509_STORE_CTX> ctx(X509_STORE_CTX_new());
+    if (!ctx || !X509_STORE_CTX_init(ctx.get(), verify_store, leaf_cert, &cert_chain) ||
+        // We need to inherit the verify parameters. These can be determined by
+        // the context: if it's a server it will verify SSL client certificates or
+        // vice versa.
+        !X509_STORE_CTX_set_default(ctx.get(), is_server ? "ssl_client" : "ssl_server") ||
+        // Anything non-default in "param" should overwrite anything in the ctx.
+        !X509_VERIFY_PARAM_set1(X509_STORE_CTX_get0_param(ctx.get()),
+                                SSL_CTX_get0_param(ssl_ctx)) ||
+        // ClientHelloOuter connections use a different name.
+        (!ech_name_override.empty() &&
+         !X509_VERIFY_PARAM_set1_host(X509_STORE_CTX_get0_param(ctx.get()),
+                                      ech_name_override.data(), ech_name_override.size()))) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_X509_LIB);
+      if (ssl_extended_info) {
+        ssl_extended_info->setCertificateValidationStatus(
+            Envoy::Ssl::ClientValidationStatus::Failed);
+      }
+      stats_.fail_verify_error_.inc();
+      std::string error("verify cert failed: init and setup X509_STORE_CTX");
+      if (error_details != nullptr) {
+        *error_details = error;
+      }
+      ENVOY_LOG(debug, error);
+      return Ssl::ValidateResult::Failed;
+    }
+    int ret = X509_verify_cert(ctx.get());
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(
+          ret == 1 ? Envoy::Ssl::ClientValidationStatus::Validated
+                   : Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+
+    if (ret <= 0) {
+      stats_.fail_verify_error_.inc();
+      std::string error =
+          absl::StrCat("verify cert failed: ", Utility::getX509VerificationErrorInfo(ctx.get()));
+      if (out_alert != nullptr) {
+        *out_alert = SSL_alert_from_verify_result(X509_STORE_CTX_get_error(ctx.get()));
+      }
+      if (error_details != nullptr) {
+        *error_details = error;
+      }
+      ENVOY_LOG(debug, error);
+      return allow_untrusted_certificate_ ? Ssl::ValidateResult::Successful
+                                          : Ssl::ValidateResult::Failed;
+    }
+  }
```

Comment:
This method adds quite a bit of new logic which looks reasonable to me, but I'm not sure if there are any pitfalls here. Is this code based on an existing implementation somewhere? If so can you point me at it? (Maybe we could share code?) If not, we should probably have an SSL expert take a look at this (davidben) once we're otherwise happy with the PR. Does that sound reasonable?